components/browser/state/src/main/java/mozilla/components/browser/state/reducer/DownloadStateReducer.kt

```diff
@@ -0,0 +1,48 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+package mozilla.components.browser.state.reducer
+
+import mozilla.components.browser.state.action.DownloadAction
+import mozilla.components.browser.state.state.BrowserState
+import mozilla.components.browser.state.state.content.DownloadState
+import mozilla.components.browser.state.state.content.DownloadStatus
+
+internal object DownloadStateReducer {
+
+    /**
+     * [DownloadAction] Reducer function for modifying [BrowserState.queuedDownloads].
+     */
+    fun reduce(state: BrowserState, action: DownloadAction): BrowserState = when (action) {
+        is DownloadAction.DownloadStartedAction -> {
+            state.changeDownloadStatusOf(downloadState = action.download, newState = DownloadStatus.RUNNING)
+        }
+        is DownloadAction.AddDownloadAction -> {
+            state.changeDownloadStatusOf(downloadState = action.download, newState = DownloadStatus.QUEUED)
+        }
+        is DownloadAction.DownloadCompletedAction -> {
+            state.changeDownloadStatusOf(downloadState = action.download, newState = DownloadStatus.COMPLETED)
+        }
+        is DownloadAction.DownloadFailedAction -> {
+            state.changeDownloadStatusOf(downloadState = action.download, newState = DownloadStatus.FAILED)
+        }
+    }
```

Comment:
I still think we also need an action to "consume" a download here once we no longer need it in the state?